src/rust/engine/process_execution/sandboxer/src/lib.rs

```diff
@@ -297,3 +321,173 @@ impl SandboxerClient {
+            store_cli_opt,
+            process: Arc::new(RwLock::new(None)),
+        })
+    }
+
+    pub fn socket_path(&self) -> &Path {
+        self.socket_path.as_ref()
+    }
+
+    pub async fn is_alive(&self) -> Result<bool, String> {
+        let mut locked_proc = self.process.write().await;
+        if let Some(proc) = locked_proc.as_mut() {
+            proc.process.check_child_has_exited().map(|b| !b)
+        } else {
+            Ok(false)
+        }
+    }
+
+    pub async fn spawn(&self) -> Result<(), String> {
+        let mut locked_proc = self.process.write().await;
+        self.do_spawn(&mut locked_proc).await
+    }
+
+    async fn do_spawn(
+        &self,
+        locked_proc: &mut WriteLockedSandboxerProcess<'_>,
+    ) -> Result<(), String> {
+        ensure_parent_dir(&self.log_path)?;
+        let logfile = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(&self.log_path)
+            .await
+            .map_err(|e| e.to_string())?;
+        let mut cmd = Command::new(&self.sandboxer_bin);
+        cmd.arg("--socket-path");
+        cmd.arg(self.socket_path.as_os_str());
+        cmd.args(self.store_cli_opt.to_cli_args());
+        // Pass the calling code's max log level to the sandboxer process.
+        cmd.env("RUST_LOG", log::max_level().as_str());
+        cmd.stderr(Stdio::from(logfile.into_std().await));
+
+        debug!(
+            "Spawning sandboxer with cmd: {} {}",
+            &self.sandboxer_bin.to_string_lossy(),
+            cmd.as_std()
+                .get_args()
+                .map(|a| a.to_string_lossy())
+                .collect::<Vec<_>>()
+                .join(" ")
+        );
+
+        if locked_proc.is_none() {
+            let process = ManagedChild::spawn(&mut cmd, Some(Duration::from_millis(1000)))
+                .map_err(|e| e.to_string())?;
+            debug!("Spawned sandboxer process {:?}", process.id());
+            let client = SandboxerClient::connect_with_retries(&self.socket_path).await?;
+            let _ = locked_proc.insert(SandboxerProcess { process, client });
+        }
+        Ok(())
+    }
+
+    async fn kill(&self) {
+        let mut locked_proc = self.process.write().await;
+        self.do_kill(&mut locked_proc).await
+    }
+
+    async fn do_kill(&self, locked_proc: &mut RwLockWriteGuard<'_, Option<SandboxerProcess>>) {
+        // Note that we don't want to remove the socket file here, as it may no longer
+        // be under the control of the sandboxer process we're killing.
+        if let Some(ref mut proc) = **locked_proc {
+            // Best effort to shut down, ignoring any errors (which would
+            // typically be due to the process already being dead).
+            let _ = proc.process.attempt_shutdown_sync();
+        }
+        locked_proc.take();
+    }
+
+    async fn client(&self) -> Result<SandboxerClient, String> {
+        {
+            let proc = self.process.read().await;
```

Comment:
Should this code check whether the process is still running? Or should there be a way to detect the subprocess exiting early? (general comment, such checks need not be in this function)

Reply:
It's unnecessary since the caller of `client()` - `materialize_directory()` - has retry logic already. And it is more efficient to optimistically try and fail and retry than to pessimistically check process status each time. Also, the retry logic handles the "process is running but not responding" case, which a check here would not.